wine-wayland: added derivation for building the experimental Wayland driver for Wine

jmc-figueira opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Added the necessary derivations and aliases to build the experimental Wayland driver for Wine (from Collabora: https://gitlab.collabora.com/alf/wine/-/tree/wayland). The derivation was based off of the existing derivation for the staging version, as well as the derivation present in this user's flake: https://github.com/cidkidnix/nixcfg/blob/master/flake.nix#L67

This builds the wayland-compatible version for all Wine builds, with a certain caveat still to be solved (hence the draft), with which I request some assistance (more information follows below).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
maintainers: add jmc-figueira

Commit:
wine-wayland: added derivation for building the experimental Wayland driver for Wine

jmc-figueira commented:
The main issue encountered seems to be related to the `wineWow` builds. Namely, the build fails with the following errors:

```
error: builder for '/nix/store/ax5iacw1xxarpadf3zrhy5wmi38xd480-wine-wow-7.0-rc2-wayland.drv' failed with exit code 1;
       last 10 log lines:
       > creating Makefile rules... done
       > configure: creating ./config.status
       > config.status: creating include/config.h
       > config.status: linking /build/source/tools/winewrapper to wine
       > config.status: linking /build/source/tools/winewrapper to wine64
       > config.status: executing include/stamp-h commands
       > config.status: executing tools/makedep commands
       > config.status: executing Makefile commands
       > makedep: error: open /stable/viewporter/viewporter.xml : No such file or directory
       > config.status: error: could not create Makefile
       For full logs, run 'nix log /nix/store/ax5iacw1xxarpadf3zrhy5wmi38xd480-wine-wow-7.0-rc2-wayland.drv'.
```

From what I could gather, `viewporter.xml` is provided by the `wayland-protocols` package, which is included as a `buildInput` for all appropriate architectures. Note that this error does not occur on neither `wine` nor `wine64` builds. My assumption thus far is that this is caused by the need to link both architectures for the `wineWow` build, but I have not been able to figure out how to solve this issue.

Commit:
wine-wayland: fixed wineWow builds with wayland support

jmc-figueira commented:
I believe I have managed to solve the aforementioned issue regarding `wineWow` builds. Please review this PR, as there may be some remaining issues with the way I went about making this derivation, it might not be the optimal way. Regardless, I will post the `nixpkgs-review` and test results as soon as they are finished (however, manually building and testing has been successful).

FlorianFranzen commented on pkgs/misc/emulators/wine/wayland.nix:
```diff
@@ -0,0 +1,30 @@
+  configureFlags = old.configureFlags
+    ++ [ "--with-wayland" ]
+    ++ lib.optionals vulkanSupport [ "--with-vulkan" ]
+    ++ lib.optionals vkd3dSupport [ "--with-vkd3d" ];
```
These three overrides probably should just be moved to `base.nix` under a wayland feature flag. Then this file can be deleted.

FlorianFranzen commented on pkgs/misc/emulators/wine/default.nix:
```diff
@@ -67,4 +69,13 @@ in if wineRelease == "staging" then
+
+      pkgArches = lib.optionals (wineBuild == "wine32" || wineBuild == "wineWow") [ pkgsi686Linux ] ++ lib.optionals (wineBuild == "wine64" || wineBuild == "wineWow") [ pkgs ];
+    }
+    else
```
This indirection here is indeed weird. I know you only followed what staging does, but I would probably just toggle a wayland feature instead and do all the work in `base.nix`. Then `packages.nix` will provide all this information for you.

FlorianFranzen reviewed (changes requested):
Nice work. I think however if you move the overrides from `wayland.nix` into `base.nix` your patch will be much cleaner and a lot more future proof. 

I imagine wayland support will make it into other variants some day, so we probably just want a new entry in  `supportFlags` for it.

jmc-figueira commented:
> Nice work. I think however if you move the overrides from `wayland.nix` into `base.nix` your patch will be much cleaner and a lot more future proof.
> 
> I imagine wayland support will make it into other variants some day, so we probably just want a new entry in `supportFlags` for it.

I did try that beforehand, the reason I decided against it was that it caused massive rebuilds, for every wine variant and every package which has wine as a dependency, regardless of the variant. You can see this through the flags that ofborg added before today's force push (which amended this). I decided to separate it in this fashion to reduce that. If you believe that moving the changes to `base.nix` would be better for the reasons you specified, I can go back to that approach.

jmc-figueira commented on pkgs/misc/emulators/wine/builder-wow.sh:
```diff
@@ -11,14 +12,14 @@ mkdir -p $TMP/wine-wow $TMP/wine64
 cd $TMP/wine64
 sourceRoot=`pwd`
-configureFlags="--enable-win64"
+configureFlags="${preFlags} --enable-win64"
```
Originally I did use `+=`; this, however, caused the `configureFlags` for the 32-bit build to keep the `--enable-win64` flag, which is not intended, so I decided to separate it into `preFlags`, which contains `--with-wayland`, etc. and simply add the flags necessary for `wow` to that.

Commit:
wine-wayland: fixed vkd3d dependency in wineWow builds and standardized the derivation

As recommended by @FlorianFranzen, the derivation for wine-wayland builds was moved from a separate "patch-like" derivation to base.nix, by setting the appropriate supportFlags. This was also done to solve an issue when building wineWow builds, where they would fail due to the 32-bit vkd3d not being linked appropriately.

jmc-figueira commented:
I ended up following @FlorianFranzen's suggestion anyways, as I needed a way to work around the following issue, where the 32-bit version of the `vkd3d` library wouldn't get linked appropriately, as it was being added internally through `callPackage`. Since this change would cause the aforementioned rebuilds anyways, I followed through with the rest of the suggestions as well.

```
checking for -lvkd3d... not found
checking for -lvkd3d-shader... not found
configure: error: vkd3d 32-bit development files not found (or too old), Direct3D 12 won't be supported.
This is an error since --with-vkd3d was requested.
error: builder for '/nix/store/iflxwk0ab0b5vf74yjwhhxkwqbhq4vjm-wine-wow-7.0-rc2-wayland.drv' failed with exit code 1
```

Commit:
wine-wayland: disabled Xinerama support for Wayland builds, removing needless dependencies

lovesegfault commented:
Result of `nixpkgs-review pr 153932` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>lutris</li>
    <li>lutris-free</li>
    <li>lutris-unwrapped</li>
  </ul>
</details>
<details>
  <summary>22 packages failed to build:</summary>
  <ul>
    <li>bottles</li>
    <li>grapejuice</li>
    <li>pipelight</li>
    <li>playonlinux</li>
    <li>q4wine</li>
    <li>winbox</li>
    <li>wine (winePackages.full ,winePackages.stableFull ,winePackages.unstableFull)</li>
    <li>wine64Packages.base (wine64Packages.stable ,wine64Packages.unstable)</li>
    <li>wine64Packages.minimal</li>
    <li>wine64Packages.stagingFull</li>
    <li>wine64Packages.wayland</li>
    <li>wine64Packages.waylandFull</li>
    <li>winePackages.minimal</li>
    <li>winePackages.staging</li>
    <li>wineWowPackages.base (wineWowPackages.stable ,wineWowPackages.unstable)</li>
    <li>wineWowPackages.minimal</li>
    <li>wineWowPackages.staging</li>
    <li>wineWowPackages.stagingFull</li>
    <li>wineWowPackages.wayland</li>
    <li>wineWowPackages.waylandFull</li>
    <li>yabridge</li>
    <li>yabridgectl</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>wine-staging (winePackages.stagingFull)</li>
    <li>wine-wayland (winePackages.waylandFull)</li>
    <li>wine64 (wine64Packages.full ,wine64Packages.stableFull ,wine64Packages.unstableFull)</li>
    <li>wine64Packages.staging</li>
    <li>winePackages.base (winePackages.stable ,winePackages.unstable)</li>
    <li>winePackages.wayland</li>
    <li>wineWowPackages.full (wineWowPackages.stableFull ,wineWowPackages.unstableFull)</li>
  </ul>
</details>

jmc-figueira commented:
Those build failures are rather strange; could you share the reason for them?

lovesegfault commented:
> Those build failures are rather strange; could you share the reason for them?

If I had to guess my machine OOM'd :P

fufexan commented:
Result of `nixpkgs-review pr 153932` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>lutris</li>
    <li>lutris-free</li>
    <li>lutris-unwrapped</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>winePackages.wayland</li>
    <li>wineWowPackages.waylandFull</li>
  </ul>
</details>
<details>
  <summary>27 packages built:</summary>
  <ul>
    <li>bottles</li>
    <li>grapejuice</li>
    <li>pipelight</li>
    <li>playonlinux</li>
    <li>q4wine</li>
    <li>winbox</li>
    <li>wine (winePackages.full ,winePackages.stableFull ,winePackages.unstableFull)</li>
    <li>wine-staging (winePackages.stagingFull)</li>
    <li>wine-wayland (winePackages.waylandFull)</li>
    <li>wine64 (wine64Packages.full ,wine64Packages.stableFull ,wine64Packages.unstableFull)</li>
    <li>wine64Packages.base (wine64Packages.stable ,wine64Packages.unstable)</li>
    <li>wine64Packages.minimal</li>
    <li>wine64Packages.staging</li>
    <li>wine64Packages.stagingFull</li>
    <li>wine64Packages.wayland</li>
    <li>wine64Packages.waylandFull</li>
    <li>winePackages.base (winePackages.stable ,winePackages.unstable)</li>
    <li>winePackages.minimal</li>
    <li>winePackages.staging</li>
    <li>wineWowPackages.base (wineWowPackages.stable ,wineWowPackages.unstable)</li>
    <li>wineWowPackages.full (wineWowPackages.stableFull ,wineWowPackages.unstableFull)</li>
    <li>wineWowPackages.minimal</li>
    <li>wineWowPackages.staging</li>
    <li>wineWowPackages.stagingFull</li>
    <li>wineWowPackages.wayland</li>
    <li>yabridge</li>
    <li>yabridgectl</li>
  </ul>
</details>

Commit:
Merge branch 'master' into wine-wayland

Commit:
Fixed typo while fixing maintainer list merge conflict

lovesegfault commented:
@GrahamcOfBorg eval

jmc-figueira commented:
> > Those build failures are rather strange; could you share the reason for them?
> 
> If I had to guess my machine OOM'd :P

Something similar happens to me with `nixpkgs-review`, where I run out of "disk space" (due to having a tmpfs root, most likely, which means it's actually OOMing). However, building the packages individually works and I can confirmed I have overridden Lutris to use `wineWowPackages.waylandFull` and it works as intended.

As for Lutris being listed as broken, all I could find was this: https://github.com/NixOS/nixpkgs/pull/155529#issuecomment-1019567947

which suggests it's an unrelated issue to Wine.

lovesegfault commented:
Result of `nixpkgs-review pr 153932` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>airwave</li>
  </ul>
</details>
<details>
  <summary>32 packages built:</summary>
  <ul>
    <li>bottles</li>
    <li>grapejuice</li>
    <li>lutris</li>
    <li>lutris-free</li>
    <li>lutris-unwrapped</li>
    <li>pipelight</li>
    <li>playonlinux</li>
    <li>q4wine</li>
    <li>winbox</li>
    <li>wine (winePackages.stableFull ,winePackages.unstableFull)</li>
    <li>wine-staging (winePackages.stagingFull)</li>
    <li>wine-wayland (winePackages.waylandFull)</li>
    <li>wine64 (wine64Packages.stableFull ,wine64Packages.unstableFull)</li>
    <li>wine64Packages.base (wine64Packages.stable ,wine64Packages.unstable)</li>
    <li>wine64Packages.minimal</li>
    <li>wine64Packages.staging</li>
    <li>wine64Packages.stagingFull</li>
    <li>wine64Packages.wayland</li>
    <li>wine64Packages.waylandFull</li>
    <li>winePackages.base (winePackages.stable ,winePackages.unstable)</li>
    <li>winePackages.minimal</li>
    <li>winePackages.staging</li>
    <li>winePackages.wayland</li>
    <li>wineWowPackages.base (wineWowPackages.stable ,wineWowPackages.unstable)</li>
    <li>wineWowPackages.full (wineWowPackages.stableFull ,wineWowPackages.unstableFull)</li>
    <li>wineWowPackages.minimal</li>
    <li>wineWowPackages.staging</li>
    <li>wineWowPackages.stagingFull</li>
    <li>wineWowPackages.wayland</li>
    <li>wineWowPackages.waylandFull</li>
    <li>yabridge</li>
    <li>yabridgectl</li>
  </ul>
</details>

lovesegfault commented:
Seems like this is ready to be merged?

lovesegfault commented:
@GrahamcOfBorg eval
